python/pyspark/sql/tests.py

```diff
@@ -3166,6 +3166,88 @@ def test_filtered_frame(self):
+
+        udf = pandas_udf(lambda x: x, DoubleType())
+        self.assertEquals(udf.returnType, DoubleType())
+        self.assertEquals(udf.evalType, PythonEvalType.PANDAS_SCALAR_UDF)
+
+        udf = pandas_udf(lambda x: x, DoubleType(), PandasUDFType.SCALAR)
+        self.assertEquals(udf.returnType, DoubleType())
+        self.assertEquals(udf.evalType, PythonEvalType.PANDAS_SCALAR_UDF)
+
+        udf = pandas_udf(lambda x: x, 'v double', PandasUDFType.GROUP_MAP)
+        self.assertEquals(udf.returnType, StructType([StructField("v", DoubleType())]))
+        self.assertEquals(udf.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+        udf = pandas_udf(lambda x: x, 'v double',
+                         functionType=PandasUDFType.GROUP_MAP)
+        self.assertEquals(udf.returnType, StructType([StructField("v", DoubleType())]))
+        self.assertEquals(udf.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+        udf = pandas_udf(lambda x: x, returnType='v double',
+                         functionType=PandasUDFType.GROUP_MAP)
+        self.assertEquals(udf.returnType, StructType([StructField("v", DoubleType())]))
+        self.assertEquals(udf.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+    def test_pandas_udf_decorator(self):
+        from pyspark.rdd import PythonEvalType
+        from pyspark.sql.functions import pandas_udf, PandasUDFType
+        from pyspark.sql.types import StructType, StructField, DoubleType
+
+        @pandas_udf(DoubleType())
+        def foo(x):
+            return x
+        self.assertEquals(foo.returnType, DoubleType())
+        self.assertEquals(foo.evalType, PythonEvalType.PANDAS_SCALAR_UDF)
+
+        @pandas_udf(returnType=DoubleType())
+        def foo(x):
+            return x
+        self.assertEquals(foo.returnType, DoubleType())
+        self.assertEquals(foo.evalType, PythonEvalType.PANDAS_SCALAR_UDF)
+
+        schema = StructType([StructField("v", DoubleType())])
+
+        @pandas_udf(schema, PandasUDFType.GROUP_MAP)
+        def foo(x):
+            return x
+        self.assertEquals(foo.returnType, schema)
+        self.assertEquals(foo.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+        @pandas_udf(schema, functionType=PandasUDFType.GROUP_MAP)
+        def foo(x):
+            return x
+        self.assertEquals(foo.returnType, schema)
+        self.assertEquals(foo.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+        @pandas_udf(returnType=schema, functionType=PandasUDFType.GROUP_MAP)
+        def foo(x):
+            return x
+        self.assertEquals(foo.returnType, schema)
+        self.assertEquals(foo.evalType, PythonEvalType.PANDAS_GROUP_MAP_UDF)
+
+    def test_udf_wrong_arg(self):
+        from pyspark.sql.functions import pandas_udf, PandasUDFType
+
+        with QuietTest(self.sc):
+            with self.assertRaisesRegexp(ValueError, 'return type'):
+                @pandas_udf(PandasUDFType.GROUP_MAP)
```

Comment:
a little insane but how about `@pandas_udf(returnType=PandasUDFType.GROUP_MAP)`?

Reply:
Not insane at all :) Added the test.